src/Core/Event.js

```diff
@@ -0,0 +1,28 @@
+/**
+ * @param {Event}
+ */
+function handleEvent (event) {
+	try {
+		var type = event.type
+		var element = this
+		var host = element.host
+		var instance = host.instance
+		var callback = element.event[type]
+		var state
+
+		if (!callback)
+			return
+
+		if (typeof callback === 'function')
+			state = callback.call(instance, event)
```

Comment:
Maybe we can pass state and props to event listeners. This will allow stateful function components to also use arrow functions as event handlers. The only point i'm trying to figure out is in which order, `e, props, state` vs `e, state, props`. The second pattern looks like it would be ideal for the most common use case of the arguments mentioned.

```js
// either
const handleEvent = (e, props, state) => ({active: state.active, size: props.size})
// or
const handleEvent = (e, state, state) => ({active: state.active, size: props.size})

const Button = (props, state, context) => h('button', {...}, props.children)
```

Reply:
or 
```js
const handleEvent = (e, {props, state}) => ({active: state.active, size: props.size})
```
I'm not sure I'd pass extra arguments to event handlers though

Reply:
The why reasoning behind this thought was that since we can write stateful function components, for example:

```js
const X = (props, {value, index, ...state}) => {
	return [
		h('input', {value: value, onInput: handleInput}),
		h('h1', index)
	]
}

X.getInitialState = ({index, value, ...props}) => {
	return {value: value, index: index}
}

dio.render(h(X, {index: 0, value: ''}))
```

How could we then right the `handleInput` function in a way that we can update the components state using the internal state & props of the component.

On one hand we could right it as a function that preserves the`this` context, and access state and props through `this.props` or `this.state` when/where needed.

```js
function handleInput (e) {
  return {value: e.currentTarget.value, this.state.index: this.state.index+1}
  // we could just as well directly call `this.setState` with the object.
}
```

On the other hand JavaScript now has arrow functions that do not preserve the functions `this` context. In this case we can't do the described unless we have arguments that give us references to the components internal state & props.

```js
const handleInput = (e, props, {index, ...state}) => {
  return {value: e.currentTarget.value, index: index+1}
}
```

This would allow the use of the second pattern(arrow functions) to retain a path to updating the state using the components current internal state/props.

Reply:
I'm a fan of partial application (for function components)

```js
h('input', {oninput: _.partialRight(handleInput, props)})
handleInput = (e, props) => {}
```

or simply inline

```js
h('input', {oninput: (e) => handleInput(e, props)})
```

I guess I prefer less framework, more functional paradigms. But I can see an argument for both

Reply:
I imagine this should make functional paradigms more natural and less expensive, relegating the need to create new functions within every `render` context in order to preserve `props` & `state` references.